Standardize ProviderConfig for Cluster and Machine

craigtracey opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
In order to maintain type parity for providerConfig across both Machine
and Cluster types, change the ClusterSpec.ProviderConfig to the
ProviderConfig type.

Additionally, as all changes to the API require regenerating code, add a
Makefile that will simplify that process.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```
None
```

<!-- All reviews default to cc'ing the kube-deploy-reviewers github group. -->
@kubernetes/kube-deploy-reviewers

craigtracey committed:
Standardize ProviderConfig for Cluster and Machine

In order to maintain type parity for providerConfig across both Machine
and Cluster types, change the ClusterSpec.ProviderConfig to the
ProviderConfig type.

Additionally, as all changes to the API require regenerating code, add a
Makefile that will simplify that process.

Signed-off-by: Craig Tracey <craigtracey@gmail.com>

kris-nova commented:
/ok-to-test

krousey commented on cluster-api/pkg/apis/cluster/v1alpha1/common_types.go:
```diff
@@ -0,0 +1,43 @@
+
+	// No more than one of the following may be specified.
+
+	// Value is an inlined, serialized representation of the node
```
remove "node" and replace appropriately.

krousey commented on cluster-api/pkg/apis/cluster/v1alpha1/common_types.go:
```diff
@@ -0,0 +1,43 @@
+
+import runtime "k8s.io/apimachinery/pkg/runtime"
+
+// ProviderConfig defines the configuration to use during node creation.
```
"node"

krousey commented on cluster-api/Makefile:
```diff
@@ -0,0 +1,25 @@
+all: genapi genconversion
+
+genapi:
+	go get -u github.com/kubernetes-incubator/apiserver-builder/cmd/apiregister-gen
```
This means we're always using the tool at head. Is it possible to vendor, build from vendor, and use that copy?

krousey commented on cluster-api/Makefile:
```diff
@@ -0,0 +1,25 @@
+	apiregister-gen -i ./pkg/apis,./pkg/apis/cluster/v1alpha1
+
+genconversion:
+	go get -u k8s.io/code-generator/cmd/conversion-gen
```
Same vendoring comment for this.

krousey reviewed (changes requested):
Thanks for doing this. Just a couple of comments.

craigtracey committed:
Update Makefile to lock api generation utilities

Modify the dep config to lock our codegen dependencies and make use of
these dependencies in a top-level Makefile.

Signed-off-by: Craig Tracey <craigtracey@gmail.com>

craigtracey committed:
Update vendored dependencies

Updating vendor dependencies in response to introduction of
k8s.io/code-generator.

Signed-off-by: Craig Tracey <craigtracey@gmail.com>

craigtracey committed:
Update comments to reflect common type adjustment

As this code has moved from being Machine-specific to being generic,
remove references to "node" in favor of "resource." "resource" has been
chosen as this type may be leveraged by any type that would like to
implement "providerConfig" semantics.

Signed-off-by: Craig Tracey <craigtracey@gmail.com>

krousey commented:
/lgtm
